### PR TITLE
Copter/Plane/Rover: Add a bitmask parameter to block mode entry from GCS

### PR DESCRIPTION
**Pull request by IamPete1**
This adds a new parameter to block mode entry from GCS. This does not block mode entry from any other source. The idea is to prevent GCS mistakes from putting the vehicle into a mode that the pilot is not experienced with or expecting. For example if the GCS operator mean to hit alt-hold but accidentally selected acro. 

image

This does not prevent entry from any other source, the main other candidate would be RC, but in that case you can just not setup switches for modes you don't want to use. 

There are some modes that it is not allowed to disable, stabilize, alt-hold RTL and land. Since this is only GCS I can see the case for also allowing disabling stabilize and alt-hold, switching into them without coordination with the pilot will also cause issues. 

Unfortunately there are too many modes to map over to the bitmask using the existing mode numbers, so I have had to re-map using a switch case. 

This is just copter, but if we think this is worth while I think it makes sense to do the other vehicles too.

This work is sponsored by ARACE.

### CODE REVIEW

**Comment by pompecukor** [comment]
@IamPete1 Excellent, exactly what we wanted and would definitely appreciate for Plane too, if this one gets a nod from the "big guys" ;) . Thank you.

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -1224,6 +1224,34 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
     AP_GROUPINFO("TKOFF_RPM_MAX", 7, ParametersG2, takeoff_rpm_max, 0),
 #endif
 
+    // @Param: GCS_MODE_BLOCK
```
Comment (rmackay9) [inline ArduCopter/Parameters.cpp:1227]: Maybe we could call the parameter FLTMODE_DISABLE(D) and then apply it to all situations?  E.g. just completely block the vehicle from being able to enter these modes.

Comment (IamPete1) [reply ArduCopter/Parameters.cpp:1227]: I think all situations is a bad idea and I don't really see the benefit. There are no other paths to accidentally change to the wrong mode. Via RC requires you to have setup the modes you want on the switch, scripting requires you to have added a script that does the wrong thing. Then that just leaves all the various failsafe reasons for changing modes, I can't see any case where blocking a failsafe mode change is a good idea. 

TLDR: blocking all modes changes leaves the possibility for conflicting parameters, EG `RCx_OPTION` set to mode loiter but loiter disabled.

**Comment by khancyr** [comment]
IMHO, we should be able to block all modes
Easy mistake is to fly without rc in guided or auto and then, switch to stabilize or alt_hold .... instant crash

File: ArduCopter/Parameters.cpp (hunk 1 above)
Comment (pompecukor) [reply ArduCopter/Parameters.cpp:1227]: I agree with Peter, from use case perspective. This is to prevent pilot mistakes. There is not other path where it would be a mistake.
Flight modes configured on RC will all be intentional and planned to use. Otherwise it would not be configured on the RC. Then only other ways to change modes is mission or fail-safes and scripts. 
Modes setup on FS or scripts we would not want to block obviously.

**Comment by pompecukor** [comment]
> 

But the opposite to that argument is to have RC and an emergency and the only way to land it is Althold, then again instant crash.
I don't know how many people fly entirely without any RC for redundancy. All the many hundred drones that we deploy all have RC. I mean just think landing repositioning. You need RC.

**Comment by IamPete1** [comment]
> IMHO, we should be able to block all modes Easy mistake is to fly without rc in guided or auto and then, switch to stabilize or alt_hold .... instant crash

Yeah, I think your right, because this is GCS only we should be allowed to bock alt hold and stabilize as well. That would just leave RTL and land as the only modes your not allowed to disable. I can see the case for completeness that you should be able to block them too.

**Comment by Davidsastresas** [comment]
This is great stuff, we can pick this parameter from GCS to not display non desired flight modes. Thanks!

**Comment by pompecukor** [comment]
> This is great stuff, we can pick this parameter from GCS to not display non desired flight modes. Thanks!

That would be great also!

**Comment by IamPete1** [comment]
Rebased and added stabilize and alt hold as block-able modes. The only remaining none-block-able modes are RTL and Land.

**Comment by IamPete1** [comment]
The addition of those modes now means were one bit too long for float parameter transport at 24 bits.

**Comment by rmackay9** [comment]
So I think the main question remains whether modes should be completely blocked or only blocked when requested via a MAVLink command.

Can we think of a reasonable case where we would want to make the mode available on an RC switch but not via a MAVlink command?

I think the answer depends upon whether you think this parameter is defining modes that the vehicle simply shouldn't be flown in vs whether it is a narrower scope parameter to avoid the very specific case of a mistaken GCS button press.

The most common ways modes can be changed is probably:

1. MAVLink (including GCS and companion computers)
2. RC (flight mode switch, aux switch)
3. Failsafes
4. Lua scripts

... but if you want to confuse yourself, here's our full [list of mode change reasons](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Vehicle/ModeReason.h)

**Comment by khancyr** [comment]
"Can we think of a reasonable case where we would want to make the mode available on an RC switch but not via a MAVlink command?"
We could argue that if you have RC you may want to keep stabilize and alt_hold. But that still leave the option for crash if you are out of range or are doing something else during the switch.
We could probably use a LUA script to do the bitmask tweaking in flight if we want more granualirity depending on flight phase or conditions.

@pompecukor "I don't know how many people fly entirely without any RC for redundancy, but I am very confident it is a single digit percentage." 
Don't forget all those users that are out of range, don't have correctly configured RC, flying on wifi, etc. You can check issue list of discuss to see numerous crash from GUIDED -> Loiter transition because people forget about putting some throttle. So from a use case point of view, blocking this even with RC could be a good things

**Comment by rmackay9** [comment]
@khancyr @pompecukor,

By the way, I often see misinterpretation of "without RC" so I'd like to clarify that "without RC" also means "without a Joystick" because a Joystick connected to a ground station still qualifies as an RC controller at least from the vehicle code point of view.

So the number of operators with no traditional RC (e.g. a Futaba transmitter and receiver) may be significant and growing, the number of operations with no RC at all (e.g. also no Joystick) is very small.  I can't really imagine why an operator would choose to have no RC at all.  .. this is a bit off topic but..

**Comment by rmackay9** [comment]
@khancyr, yes indeed, a drone show in particular is a perfect example of where no RC could serious be considered.  txs

**Comment by IamPete1** [comment]
Being only GCS is because users already have the option to prevent unwanted mode switch from RC. They can simply not assign the mode the don't want to use to a RC switch. Yes that is a little more complex than just disabling it in the bitmask. But the key thing I wanted to avoid was a parameter conflict. What if the users has set a RC switch to switch into loiter (for example) and then disabled loiter with this new parameter? What should we do in that case? Which parameter takes priority? By being just GCS we avoid having to figure out edge cases like that one.

**Comment by rmackay9** [comment]
@IamPete1,

In case of a conflict between the RC switch and the new bitmask, the bitmask would certainly win.  Copter properly handles failure to enter a flight mode so it would be fine I think.

In terms of naming, it would be nice to keep the flight mode related parameters grouped as much as possible so FLTMODE_BLOCKGCS is an option..

**Comment by EosBandi** [comment]
Just for my two cents; why do we try to fix a GCS UI design error in the autopilot? Fix QGCS, disable modes there.
Also not a fan of disabling any flight modes in the flight controller, make a remote control/GCS setup that prevents the user to switch to modes that are not wanted.

**Comment by rmackay9** [comment]
@EosBandi, yes, that's a valid point and there is actually some work ongoing to make the QGC list configurable.

**Comment by IamPete1** [comment]
I agree its fundamentally a GCS problem, however its probably a vehicle property that should move with the vehicle rather than with the GCS. So the information has to live on the vehicle somewhere (does not have to be a param of course).

We have fixed GCS problems in the autopilot before, it takes a long time to roll out new GCS features, especially to handsets such as herelink.

File: ArduCopter/Parameters.cpp [hunk 2]
```diff
@@ -1224,6 +1224,36 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
...
+    // @DisplayName: Flight mode block from GCS
+    // @Description: Bitmask of flight modes to disable for GCS selection. Mode can still be accessed via RC or failsafe.
+    // @Bitmask: 0:Stabalize
```
Comment (peterbarker) [inline ArduCopter/Parameters.cpp:1230]:
```suggestion
    // @Bitmask: 0:Stabilize
```

File: ArduCopter/mode.cpp [hunk 3]
```diff
@@ -194,6 +194,133 @@ void Copter::mode_change_failed(const Mode *mode, const char *reason)
...
+{
+    // We don't have parameter space to use the mode numbers in the bitmask directly, convert to reduced range
+    uint8_t mode_bit = 0;
+    switch (mode_num) {
+        case Mode::Number::RTL:
+        case Mode::Number::LAND:
+            // These modes are always enabled
+            return true;
+
+        case Mode::Number::STABILIZE:
```
Comment (tridge) [inline ArduCopter/mode.cpp:209]: a list exposed in AP_Vehicle, where index into list is the bit would be most compact, and would allow for common code

File: ArduCopter/Parameters.cpp (hunk 1 above)
Comment (rmackay9) [reply ArduCopter/Parameters.cpp:1227]: let's call it FLTMODE_GCSBLOCK

File: ArduCopter/Parameters.cpp [hunk 4]
```diff
@@ -1224,6 +1224,36 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
...
+    // @Bitmask: 16:Smart RTL
+    // @Bitmask: 17:FlowHold
+    // @Bitmask: 18:Follow
+    // @Bitmask: 19:ZigZag
+    // @Bitmask: 20:SystemID
+    // @Bitmask: 21:Heli Autorotate
+    // @Bitmask: 22:Auto RTL
+    // @Bitmask: 23:Turtle
+    // @User: Standard
+    AP_GROUPINFO("GCS_MODE_BLOCK", 8, ParametersG2, gcs_flight_mode_disable, 0),
```
Comment (tridge) [inline ArduCopter/Parameters.cpp:1255]: FLTMODE_GCSBLOCK

File: ArduCopter/Parameters.h [hunk 5]
```diff
@@ -693,6 +693,9 @@ class ParametersG2 {
     AP_Float pldp_range_finder_minimum_m;
     AP_Float pldp_delay_s;
     AP_Float pldp_descent_speed_ms;
+
+    // Bitmask of modes to disable from gcs
+    AP_Int32 gcs_flight_mode_disable;
```
Comment (rmackay9) [inline ArduCopter/Parameters.h:698]: maybe name this something closer to the new decided parameter name.

File: ArduCopter/mode.cpp [hunk 6]
```diff
@@ -194,6 +194,133 @@ void Copter::mode_change_failed(const Mode *mode, const char *reason)
...
+
+    const bool mode_enabled = (uint32_t(g2.gcs_flight_mode_disable) & mask) == 0;
+    if (mode_enabled) {
+        return true;
+    }
+
+    // Mode disabled, try and grab a mode name to give a better warning.
+    Mode *new_flightmode = mode_from_mode_num(mode_num);
+    if (new_flightmode != nullptr) {
+        mode_change_failed(new_flightmode, "GCS entry disabled (GCS_MODE_BLOCK)");
```
Comment (rmackay9) [inline ArduCopter/mode.cpp:316]: remember to change this to match the param name too.

**Review by tridge — DISMISSED** [review]

**Comment by IamPete1** [comment]
Moved to `FLTMODE_GCSBLOCK` in AP_Vehicle and implemented in Copter, Plane and Rover.

**Comment by timtuxworth** [comment]
> I agree its fundamentally a GCS problem, however its probably a vehicle property that should move with the vehicle rather than with the GCS. So the information has to live on the vehicle somewhere (does not have to be a param of course).
> 
> We have fixed GCS problems in the autopilot before, it takes a long time to roll out new GCS features, especially to handsets such as herelink.

Yes but is a GCS+AP problem, because this is about blocking modes for a specific vehicle. The whole architecture of how AP works with GCS is that the AP is the "system of record" for any vehicle specific configurations. If you fly with quads, VTOL, fixed wing (FPV), Fixed Wing (aerobatic) .. the modes you want to disable would be different depending on which vehicle you connect to. If you connect with QGC, MP, MavProxy, APM Planner ... you would have to set the valid modes on every GCS, per vehicle. It just doesn't work.

The GCS can interpret the settings on the AP and improve the GPS (as @Davidsastresas is proposing) UI based on the vehicle specific settings sourced from the AP, but

The right place to do this is the AP(as per the PR) - the AP is the system of record for vehicle specific configurations.

**Comment by EosBandi** [comment]
> The right place to implement the restriction is the AP(as per the PR) - the AP is the system of record for vehicle-specific configurations.

Well, it can be viewed from that point. I'm not arguing. However, besides the AP part, GCS sw needs to be updated to query the block settings and change the available modes accordingly. (Michael or I can do this on Mission Planner, but others must also be updated.)
Also, I feel some kind of ambiguity about blocking modes only if mode change is initiated from GCS, but you can still change to them via RC switch or RC Override.

File: libraries/AP_Vehicle/AP_Vehicle.cpp [hunk 7]
```diff
@@ -137,6 +137,71 @@ const AP_Param::GroupInfo AP_Vehicle::var_info[] = {
     AP_SUBGROUPINFO(kdecan, "KDE_",  19, AP_Vehicle, AP_KDECAN),
 #endif
 
+#if APM_BUILD_COPTER_OR_HELI || APM_BUILD_TYPE(APM_BUILD_ArduPlane) || APM_BUILD_TYPE(APM_BUILD_Rover)
+    // @Param: FLTMODE_GCSBLOCK
+    // @DisplayName: Flight mode block from GCS
+    // @Description: Bitmask of flight modes to disable for GCS selection. Mode can still be accessed via RC or failsafe.
+    // @Bitmask{Copter}: 0:Stabilize
+    // @Bitmask{Copter}: 1:Acro
+    // @Bitmask{Copter}: 2:Alt hold
```
Comment (rmackay9) [inline libraries/AP_Vehicle/AP_Vehicle.cpp:146]: Let's make the user facing names here match what's in FLTMODE1 parameter's description.

File: ArduPlane/system.cpp [hunk 8]
```diff
@@ -225,6 +225,40 @@ static bool mode_reason_is_landing_sequence(const ModeReason reason)
 }
 #endif // AP_FENCE_ENABLED
 
+// Check if this mode can be entered from the GCS
```
Comment (peterbarker) [inline ArduPlane/system.cpp:228]:
```suggestion
#if GCS_MODE_ENTRY_FILTERING_ENABLED
// Check if this mode can be entered from the GCS
```

etc

this should also be in `GCS_Plane.cpp`

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Vehicle/AP_Vehicle.cpp (hunk 7 above)
Comment (IamPete1) [reply libraries/AP_Vehicle/AP_Vehicle.cpp]: done

File: ArduPlane/system.cpp (hunk 8 above)
Comment (IamPete1) [reply ArduPlane/system.cpp:226]: We decided on the call that it was OK where it was. Do you still want a define? I guess it should live in AP_Vehicle?

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
happy for this to be merged after my review changes are fixed

**Review by rmackay9 — APPROVED** [review]

File: ArduCopter/mode.cpp [hunk 9]
```diff
@@ -194,6 +194,52 @@ void Copter::mode_change_failed(const Mode *mode, const char *reason)
     }
 }
 
+// Check if this mode can be entered from the GCS
+bool Copter::gcs_mode_enabled(const Mode::Number mode_num)
+{
+    // List of modes that can be blocked, index is bit number in parameter bitmask
+    const uint8_t mode_list [] {
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply ArduCopter/mode.cpp]: fixed

File: ArduPlane/system.cpp [hunk 10]
```diff
@@ -225,6 +225,40 @@ static bool mode_reason_is_landing_sequence(const ModeReason reason)
...
+bool Plane::gcs_mode_enabled(const Mode::Number mode_num) const
+{
+    // List of modes that can be blocked, index is bit number in parameter bitmask
+    const uint8_t mode_list [] {
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply ArduPlane/system.cpp]: fixed

File: Rover/system.cpp [hunk 11]
```diff
@@ -221,13 +221,43 @@ void Rover::update_ahrs_flyforward()
     ahrs.set_fly_forward(flyforward);
 }
 
+// Check if this mode can be entered from the GCS
+bool Rover::gcs_mode_enabled(const Mode::Number mode_num) const
+{
+    // List of modes that can be blocked, index is bit number in parameter bitmask
+    const uint8_t mode_list [] {
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply Rover/system.cpp]: fixed

File: libraries/AP_Vehicle/AP_Vehicle.cpp [hunk 12]
```diff
@@ -137,6 +137,71 @@ const AP_Param::GroupInfo AP_Vehicle::var_info[] = {
...
+    // @Bitmask{Plane}: 6:FBWB
+    // @Bitmask{Plane}: 7:Cruise
+    // @Bitmask{Plane}: 8:Autotune
+    // @Bitmask{Plane}: 9:Auto
+    // @Bitmask{Plane}: 10:Loiter
+    // @Bitmask{Plane}: 11:Takeoff
+    // @Bitmask{Plane}: 12:Avoid ADSB
+    // @Bitmask{Plane}: 13:Guided
+    // @Bitmask{Plane}: 14:Thermal
+    // @Bitmask{Plane}: 15:Q Stabilize
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AP_Vehicle/AP_Vehicle.cpp]: Done

**Comment by rmackay9** [comment]
@IamPete1,

I think we're OK to merge this now right?
